File: Graphs/FloydWarshall.py
import random

#The object embedded in each graph node (each matrix cell)
class Edge(object):
  self.Pi #predecessor pointer
  self.Cost
  self.Label

  def __init__(self,cost):
    self.Cost = cost

#graphs are defined as square matrices (fully connected graphs) of embedded graph objects
def buildGraphMatrix(size):
  graphMatrix = []
  i = 0
  while i < size:
    graphMatrix.append([])
    i += 1
  i = 0
  while i < size:
    j = 0
    while j < size:
      if i == j:
        rando = 0
      else:
        rando = random.randint(1,10)
        # 30% probability of assigning infinity to a cell
        if rando >= 7:
          rando = 999999  #effectively infinity
      vertex = Vertex(rando)
      graphMatrix[i].append(vertex)
      j += 1
    i += 1
  return graphMatrix

def FloydWarshall(graphMatrix):
  
  for row in graphMatrix:
    for cell in graphMatrix:
      
      




graph = buildGraphMatrix(10)
printGraph(graph)
FloydWarshall(graph)
printGraph(graph)

File: Graphs/FloydWarshall.cpp
#include "stdio.h"
#include "stdlib.h"
#include "time.h"
#include "string.h"
#include <iostream>

#define MATRIX_DIM 9
#define INFINITY 99

using namespace std;

//cell of matrix (not same as vertex)
struct Cell{
  int Cost;
  int Pi;
};

void initMatrix(Cell matrix[MATRIX_DIM][MATRIX_DIM])
{
  for(int row = 0; row < MATRIX_DIM; row++){
    for(int col = 0; col < MATRIX_DIM; col++){
      //reflexive transitions get cost of zero
      if(row == col){
        matrix[row][col].Cost = 0;
        matrix[row][col].Pi = row;
      }
      else{
        //33% chance of vertex being unreachable from another: if random cost > 6 (on 1-10 scale), then assign infinity.
        int cost = rand() % 9 + 1;
        if(cost <= 6){
          matrix[row][col].Cost = cost;
        }
        else{
          matrix[row][col].Cost = INFINITY;
        }
      }
      matrix[row][col].Pi = col;
    }
  }
}


void printMatrix(Cell matrix[MATRIX_DIM][MATRIX_DIM])
{
  char buf[256] = {0};
  string rowString;

  for(int row = 0; row < MATRIX_DIM; row++){
    cout << row << ": ";
    for(int col = 0; col < MATRIX_DIM; col++){
      sprintf(buf,"{%2d : %d} ",matrix[row][col].Cost,matrix[row][col].Pi);
      rowString += buf;
      memset(buf,0,256);
    }
    cout << rowString << endl;
    rowString.clear();
  }
}


void floydWarshall(Cell matrix[MATRIX_DIM][MATRIX_DIM])
{
  
}

int main(void)
{
  //graph is represented as a square matrix
  Cell matrix[MATRIX_DIM][MATRIX_DIM];

  srand(time(NULL));

  initMatrix(matrix);
  printMatrix(matrix);

  floydWarshall(matrix);
  





  return 0;
}





